### ISSUE TITLE
handling of XDG_CACHE_DIR is broken in the cache plugin

### ISSUE THREAD

**Issue by ivg** [BinaryAnalysisPlatform/bap#1109]
1) we're trying to hijack the whole directory instead of creating a folder relative to it, which is at least not clean.
2) if a directory is a file we fail with an error:
```
$ touch cache
$ XDG_CACHE_HOME=cache bap .
Internal error: Unix.Unix_error(Unix.ENOTDIR, "rename", ".//1770a8bf-71bc-433a-a608-ff33bb3293a3/")
Backtrace:
Raised at file "src/exn.ml", line 69, characters 6-15
Called from file "bap_cache.ml", line 195, characters 2-19
Called from file "bap_cache_main.ml", line 91, characters 2-15
Called from file "bap_cache_main.ml", line 138, characters 2-27
Called from file "bap_main.ml", line 789, characters 37-46
Called from file "bap_main.ml", line 945, characters 25-28
```
3) if a directory is not empty (which the common case) we also fail, because the initialization procedure is unable to initialize the expected directory structure:
```
$ mkdir cache
$ touch cache/thing
$ XDG_CACHE_HOME=cache bap .
Internal error: (Sys_error "cache/data: No such file or directory")
Backtrace:
Raised by primitive operation at file "bap_cache_gc.ml", line 101, characters 2-18
Called from file "bap_cache_gc.ml", line 143, characters 16-43
Called from file "bap_cache_main.ml", line 138, characters 2-27
Called from file "bap_main.ml", line 789, characters 37-46
Called from file "bap_main.ml", line 945, characters 25-28
```

4) It is also assumed that `dirname p` returns a parent directory, which is not true in general, e.g.,
```
XDG_CACHE_HOME=. bap .
Internal error: Unix.Unix_error(Unix.EBUSY, "rename", ".//6aa795be-60aa-49de-9fe0-87332ea18558/")
Backtrace:
Raised at file "src/exn.ml", line 69, characters 6-15
Called from file "bap_cache.ml", line 195, characters 2-19
Called from file "bap_cache_main.ml", line 91, characters 2-15
Called from file "bap_cache_main.ml", line 138, characters 2-27
Called from file "bap_main.ml", line 789, characters 37-46
Called from file "bap_main.ml", line 945, characters 25-28
```

5) The diagnostic messages could be better, for example, when the cache directory is not writable, we should provide a descriptive error message, and maybe even indicate this as a working and disable cache.

**Comment by ivg** [comment]
it is partially fixed from the red to yellow state in #1129 points 1 and 4 are still valid issues.

**Comment by ivg** [comment]
closed by #1278
